### ISSUE TITLE
subscription returning same record multiple times incrementally

### ISSUE THREAD

**Issue by jalalat** [facebook/relay#2216]
Everything working fine on server side. On client side when we publish first comment it gets to client (react native) correctly. The second comment gets displayed two times. The third comment gets displayed three times and so on. Does it mean:

> every time comment mutation (on server side) runs it create a new subscription besides existing one. That is the only logical explanation i can think of.

I am triggering Publication of Subscription in mutatation like this:

```
outputFields: {
  commentEdge: {
     type: GraphQLCommentEdge,
     resolve: async ({ comment, postId }) => {
         // Publishing change to the COMMENT_ADDED subscription.
         await pubsub.publish(COMMENT_SUB.COMMENT_ADDED, { commentAdded: commentEdge, postId });
     }
  }
}
```

CommentSubscription.js (server side) - is very simple
```
subscribe:
   withFilter(
       () => pubsub.asyncIterator(COMMENT_SUB.COMMENT_ADDED),
       (payload, args) => payload.postId === args.postId
   ),
```

Client side: I update the store in Updater function of subscription. I commented out the Updater function of mutation thinking it might be updating the store twice. But then it does not explain the third comment three times; fourth comment four times and so on..

**Comment by gauravtiwari** [comment]
> every time comment mutation (on server side) runs it create a new subscription besides existing one. That is the only logical explanation i can think of.

Yes, sounds like it. Perhaps, consider storing existing subscriptions to a map or something and check before subscribing. The other way is to dispose a subscription and re-subscribe (but that's not ideal)

**Comment by jalalat** [comment]
Others must have same issue or everyone else is using subscription for updation and not for creation/addition. For updation it should not really matter how many times you update value (like total number of votes) and probably that's why others are not complaining?

**Comment by jalalat** [comment]
Found the solution. Problem was i was creating the edge in Mutation itself and then was passing this newly created edge to subscription. I fetched the newly created edge in subscription and just added to the store. This behavior is very similar to what anyone do in mutations. In subscription however the right way to create an edge is as mentioned in the docs:

```
const edge = ConnectionHandler.createEdge(
        store,
        notifications,
        notification,
        '<TypeOfNotificationsEdge>',
      );
```
It solved my issue.
